### PR TITLE
Add SourceText.GetContentHash

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
API approved in: https://github.com/dotnet/roslyn/issues/70752

This PR just adds the API.  IDE code to move to it is in https://github.com/dotnet/roslyn/pull/70893

### CODE REVIEW

File: src/Setup/DevDivInsertionFiles/DevDivInsertionFiles.csproj [hunk 1]
```diff
@@ -125,6 +125,7 @@
       <_Dependency Remove="System.Collections.Immutable"/>
       <_Dependency Remove="System.Diagnostics.DiagnosticSource"/>
       <_Dependency Remove="System.Drawing.Common"/>
+      <_Dependency Remove="System.IO.Hashing"/>
```
Comment (tmat) [inline src/Setup/DevDivInsertionFiles/DevDivInsertionFiles.csproj:128]: >System.IO.Hashing [](http://example.com/codeflow?start=27&length=17)

We're not using this library yet, right?

**Comment by tmat** [comment]
>     internal sealed partial record class Checksum(

Perhaps rename in a follow up to "ContentHash"?

---
Refers to: src/Workspaces/Core/Portable/Workspace/Solution/Checksum.cs:24 in 49b3e23. [](commit_id = 49b3e233e2fe40bf2cdfbd492b5617135de4a2a2, deletion_comment = False)

File: src/Setup/DevDivInsertionFiles/DevDivInsertionFiles.csproj (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Setup/DevDivInsertionFiles/DevDivInsertionFiles.csproj:128]:
```suggestion
```

**Review by tmat — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
Draft while we still work on the design.

File: src/Compilers/Core/CodeAnalysisTest/Text/SourceTextTests.cs [hunk 2]
```diff
@@ -271,7 +271,7 @@ public void ContentEqualUnaffectedByEncoding1()
             }
         }
 
-        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/70752")]
+        [Fact]
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/CodeAnalysisTest/Text/SourceTextTests.cs:274]: new logic fixes this issue.

File: src/Compilers/Core/Portable/Text/SourceText.cs [hunk 3]
```diff
@@ -598,6 +604,66 @@ public ImmutableArray<byte> GetChecksum()
...
+
+                    Span<byte> destination = stackalloc byte[128 / 8];
+                    hash.GetHashAndReset(destination);
+                    return destination.ToImmutableArray();
+                }
+                finally
+                {
+                    s_charArrayPool.Free(charBuffer);
+
+                    // Technically not needed.  But adding this reset out of paranoia.
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Text/SourceText.cs:660]:
```suggestion
                    // Technically not needed.  But adding this reset out of an abundance of paranoia.
```

**Comment by CyrusNajmabadi** [comment]
Moving out of review.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler ptal.  Thanks!

**Comment by CyrusNajmabadi** [comment]
@jaredpar for visbility and wrt https://github.com/dotnet/roslyn/issues/70878.  This takes the copy of xxhash we have and moves it to the compiler layer so it can be used for content-hashing SourceTexts (as per the offline email discussion).  So when you move us to an official package, the files would be removed from there.

File: src/Compilers/Core/CodeAnalysisTest/Text/SourceTextTests.cs [hunk 4]
```diff
@@ -285,6 +312,8 @@ public void ContentEqualUnaffectedByEncoding2()
             // This should return false, but is thrown off by the calls to GetChecksum above.  This is a bad bug that
             // will be fixed once we resolve https://github.com/dotnet/roslyn/issues/70752
             Assert.False(sourceText1.ContentEquals(sourceText2));
```
Comment (RikkiGibson) [inline src/Compilers/Core/CodeAnalysisTest/Text/SourceTextTests.cs:314]:
```suggestion
            Assert.False(sourceText1.ContentEquals(sourceText2));
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/CompilerExtensions.projitems [hunk 5]
```diff
@@ -613,7 +613,7 @@
   </ItemGroup>
   <ItemGroup>
     <Folder Include="$(MSBuildThisFileDirectory)Extensions\Compiler\" />
+    <Folder Include="$(MSBuildThisFileDirectory)Hashing\" />
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/CompilerExtensions.projitems:616]: correct

File: src/Compilers/Core/Portable/Text/SourceText.cs [hunk 6]
```diff
@@ -598,6 +604,67 @@ public ImmutableArray<byte> GetChecksum()
...
+        /// Produces a hash of this <see cref="SourceText"/> based solely on the contents it contains.  Two different
+        /// <see cref="SourceText"/> instances that are <see cref="ContentEquals"/> will have the same content hash. Two
+        /// instances of <see cref="SourceText"/> with different content are virtually certain to not have the same
+        /// hash.  This hash can be used for fingerprinting of text instances, but does not provide cryptographic
+        /// guarantees.
+        /// </summary>
+        /// <remarks>
+        /// The implementation of this function uses the XxHash128 <see
+        /// href="https://learn.microsoft.com/en-us/dotnet/api/system.io.hashing.xxhash128"/>.  However, this is not
+        /// guaranteed, and future versions of Roslyn may change this.
```
Comment (333fred) [inline src/Compilers/Core/Portable/Text/SourceText.cs:617]: I wouldn't put this in the remarks. I think the following paragraph is all we need.

**Review by 333fred — APPROVED** [review]

File: src/Compilers/Core/Portable/Text/SourceText.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Text/SourceText.cs:617]: sgtm

File: src/Compilers/Core/Portable/Text/SourceText.cs [hunk 7]
```diff
@@ -598,6 +604,67 @@ public ImmutableArray<byte> GetChecksum()
...
+        /// <para/>
+        ///
+        /// This hash is safe to use across platforms and across processes, as long as the same version of Roslyn is
+        /// used in all those locations.  As such, it is safe to use as a fast proxy for comparing text instances in
+        /// different memory spaces.  Different versions of Roslyn may produce different content hashes.
+        /// </remarks>
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Text/SourceText.cs:623]:
```suggestion
        /// <remarks>
        /// This hash is safe to use across platforms and across processes, as long as the same version of Roslyn is
        /// used in all those locations.  As such, it is safe to use as a fast proxy for comparing text instances in
        /// different memory spaces.  Different versions of Roslyn may produce different content hashes.
        /// </remarks>
```

File: src/Compilers/Core/Portable/Text/SourceText.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Text/SourceText.cs]: revised. t hanks.

**Review by RikkiGibson — APPROVED** [review]

**Pull request merged**